osquery/tables/system/windows/windows_eventlog.cpp

```diff
@@ -0,0 +1,267 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <Windows.h>
+#include <winevt.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/foreach.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <boost/property_tree/xml_parser.hpp>
+
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include <osquery/core/windows/wmi.h>
+#include <osquery/utils/conversions/join.h>
+#include <osquery/utils/conversions/split.h>
+#include <osquery/utils/conversions/windows/strings.h>
+
+namespace pt = boost::property_tree;
+
+namespace osquery {
+namespace tables {
+
+const std::string kEventLogXmlPrefix = "<QueryList><Query Id=\"0\">";
+const std::string kEventLogXmlSuffix = "</Query></QueryList>";
+
+const int kNumEventsBlock = 1024;
+
+/// Helper function to recursively parse a boost property tree
+void parseTree(const pt::ptree& tree, std::map<std::string, std::string>& res) {
+  for (const auto& node : tree) {
+    // Skip this since it's not part of the EventData.
+    if (node.first == "<xmlattr>") {
+      continue;
+    }
+
+    auto nodeName = node.second.get("<xmlattr>.Name", "");
+    if (nodeName.empty()) {
+      nodeName = node.first.empty() ? "DataElement" : node.first;
+    }
+
+    res[nodeName] = res[nodeName].empty()
+                        ? node.second.data()
+                        : res[nodeName] + "," + node.second.data();
+
+    parseTree(node.second, res);
+  }
+}
+
+void parseWelXml(pt::ptree& propTree, Row& r) {
```

Comment:
Can this be combined with parser at https://github.com/osquery/osquery/blob/master/osquery/events/windows/windowseventlogparserservice.cpp

Reply:
Thanks, @directionless for the review. It uses the same approach as `windowseventlogparserservice.cpp` and it can be combined. The log parser service has no interface function for parsing the events. I will have the changes and reuse the function for parsing the events.

Reply:
I have a strong preference for merging the two. One implementation will go out of date.

Reply:
I agree, I think we should pick one of three options before proceeding here. I'll list them in the order of my preference, but I am happy with any of the options.
1. We first land a PR that refactors the existing `parseWelXml` into a small utility library, then in this PR we use it.
2. We refactor `parseWelXml` into the utility in this PR.
3. We create an issue to track changing it in a follow up PR then promise to do it.

Reply:
Sorry for getting back to this late. I raised a PR #6567 that refactor the existing WEL parser service and get the function `processEvent` and `processEventObject`. These functions can be directly used with the on-demand table to get the attributes of the events from XML.

Reply:
I updated the PR and `parseWelXml` uses the utility functions for parsing the Windows event log that is common to both evented and non-evented table.